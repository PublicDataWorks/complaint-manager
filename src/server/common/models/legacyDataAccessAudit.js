/* 
****** This model should stay in the code base even after the
  newAuditFeature is removed 
 */

"use strict";

module.exports = (sequelize, DataTypes) => {
  const LegacyDataAccessAudit = sequelize.define(
    "legacy_data_access_audit",
    {
      id: {
        allowNull: false,
        autoIncrement: true,
        primaryKey: true,
        type: DataTypes.INTEGER
      },
      auditSubject: {
        field: "audit_subject",
        allowNull: false,
        type: DataTypes.STRING
      },
      auditDetails: {
        field: "audit_details",
        type: DataTypes.JSONB
      }
    },
    {
      tableName: "legacy_data_access_audits",
      timestamps: false
    }
  );

  LegacyDataAccessAudit.associate = models => {
    LegacyDataAccessAudit.belongsTo(models.audit, {
      as: "legacyDataAccessAudit",
      foreignKey: { name: "auditId", field: "audit_id" }
    });
  };

  return LegacyDataAccessAudit;
};
