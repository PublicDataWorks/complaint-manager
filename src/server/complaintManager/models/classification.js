"use strict";

module.exports = (sequelize, DataTypes) => {
  const Classification = sequelize.define(
    "classification",
    {
      name: {
        type: DataTypes.STRING,
        allowNull: false
      },
      message: {
        type: DataTypes.TEXT,
        allowNull: false
      }
    },
    {
      tableName: "classifications"
    }
  );

  Classification.associate = models => {
    Classification.hasMany(models.case_classification, {
      as: "caseClassification",
      foreignKey: {
        name: "classificationId",
        field: "classification_id",
        allowNull: false
      }
    });
  };

  return Classification;
};
