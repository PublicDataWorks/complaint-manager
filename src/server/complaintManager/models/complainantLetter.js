const models = "./";

module.exports = (sequelize, DataTypes) => {
  const ComplainantLetter = sequelize.define(
    "complainant_letter",
    {
      id: {
        allowNull: false,
        autoIncrement: true,
        primaryKey: true,
        type: DataTypes.INTEGER
      },
      finalPdfFilename: {
        type: DataTypes.STRING,
        field: "final_pdf_filename"
      },
      createdAt: {
        allowNull: false,
        type: DataTypes.DATE,
        field: "created_at"
      },
      updatedAt: {
        allowNull: false,
        type: DataTypes.DATE,
        field: "updated_at"
      }
    },
    {
      tableName: "complainant_letters"
    }
  );
  ComplainantLetter.associate = function(models) {
    ComplainantLetter.belongsTo(models.civilian, {
      as: "complainantCivilian",
      foreignKey: {
        name: "complainantCivilianId",
        field: "complainant_civilian_id"
      }
    });
    ComplainantLetter.belongsTo(models.case_officer, {
      as: "caseOfficers",
      foreignKey: {
        name: "complainantOfficerId",
        field: "complainant_officer_id"
      }
    });
    ComplainantLetter.belongsTo(models.cases, {
      as: "cases",
      foreignKey: {
        name: "caseId",
        field: "case_id",
        allowNull: false
      }
    });
  };

  ComplainantLetter.prototype.getCaseId = async function(transaction) {
    return this.caseId;
  };

  ComplainantLetter.prototype.modelDescription = async function(transaction) {
    return [];
  };

  ComplainantLetter.auditDataChange();
  return ComplainantLetter;
};
