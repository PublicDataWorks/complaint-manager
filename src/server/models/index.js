const fs = require("fs");
const path = require("path");
const Sequelize = require("sequelize");
const basename = path.basename(__filename);
let env = process.env.NODE_ENV || "development";
if (env === "test") {
  env += process.env.JEST_WORKER_ID;
}
const config = require(__dirname + "/../config/sequelize_config.js")[env];
const db = {};
const dataChangeAuditHooks = require("../sequelizeHooks/dataChangeAuditHooks");
const caseStatusHooks = require("../sequelizeHooks/caseStatusHooks");

let sequelize;
if (config.use_env_variable) {
  sequelize = new Sequelize(process.env[config.use_env_variable], config);
} else {
  sequelize = new Sequelize(
    config.database,
    config.username,
    config.password,
    config
  );
}

//TODO Refactor to pass in only the class references we need.
dataChangeAuditHooks.init(sequelize, Sequelize.Model);
caseStatusHooks.init(sequelize, Sequelize.Model);

fs.readdirSync(__dirname)
  .filter(file => {
    return (
      file.indexOf(".") !== 0 && file !== basename && file.slice(-3) === ".js"
    );
  })
  .forEach(file => {
    var model = sequelize["import"](path.join(__dirname, file));
    db[model.name] = model;
  });

Object.keys(db).forEach(modelName => {
  if (db[modelName].associate) {
    db[modelName].associate(db);
  }
});

db.sequelize = sequelize;
db.Sequelize = Sequelize;

module.exports = db;
