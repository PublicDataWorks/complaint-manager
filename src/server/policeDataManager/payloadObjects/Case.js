import Boom from "boom";
import { BAD_REQUEST_ERRORS } from "../../../sharedUtilities/errorMessageConstants";
import determineNextCaseStatus from "../../handlers/cases/helpers/determineNextCaseStatus";

export default class Case {
  constructor(caseModel) {
    // TODO maybe allow the constructor to get the model by id from DB (if needed)
    this._model = caseModel;
    this._status = "Initial";
  }

  get model() {
    return this._model;
  }

  get id() {
    return this._model.id;
  }

  get primaryComplainant() {
    return this._model.primaryComplainant;
  }

  get complaintType() {
    return this._model.complaintType;
  }

  get status() {
    return this._status;
  }

  setStatus = async status => {
    const nextStatus = await determineNextCaseStatus(this.status);
    if (status === nextStatus.name) {
      this._model.currentStatusId = nextStatus.id;
      this._status = nextStatus.name;
    } else if (status !== this.status) {
      throw Boom.badRequest(BAD_REQUEST_ERRORS.INVALID_CASE_STATUS);
    }
  };

  get nextStatus() {
    return determineNextCaseStatus(this.status);
  }

  get year() {
    return this._model.year;
  }

  get caseNumber() {
    return this._model.caseNumber;
  }

  get caseReferencePrefix() {
    return this._model.caseReferencePrefix;
  }

  get caseReference() {
    return this._model.caseReference;
  }

  get firstContactDate() {
    return this._model.firstContactDate;
  }

  get incidentDate() {
    return this._model.incidentDate;
  }

  get intakeSourceId() {
    return this._model.intakeSourceId;
  }

  get districtId() {
    return this._model.districtId;
  }

  get incidentTime() {
    return this._model.incidentTime;
  }

  get incidentTimezone() {
    return this._model.incidentTimezone;
  }

  get narrativeSummary() {
    return this._model.narrativeSummary;
  }

  get narrativeDetails() {
    return this._model.narrativeDetails;
  }

  get pibCaseNumber() {
    return this._model.pibCaseNumber;
  }

  get createdBy() {
    return this._model.createdBy;
  }

  get assignedTo() {
    return this._model.assignedTo;
  }
}
