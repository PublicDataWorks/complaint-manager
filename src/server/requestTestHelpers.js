import jwt from "jsonwebtoken";
import fs from "fs";
import path from "path";

const config = require("./config/config")[process.env.NODE_ENV];

function buildTokenWithPermissions(permissions, nickname) {
  const privateKeyPath = path.join(__dirname, "config", "test", "private.pem");
  const cert = fs.readFileSync(privateKeyPath);

  const payload = {
    foo: "bar",
    scope: `${config.authentication.scope} ${permissions}`
  };
  payload[`${config.authentication.nicknameKey}`] = nickname;

  const options = {
    audience: config.authentication.audience,
    issuer: config.authentication.issuer,
    algorithm: config.authentication.algorithm
  };

  return jwt.sign(payload, cert, options);
}

module.exports = buildTokenWithPermissions;
