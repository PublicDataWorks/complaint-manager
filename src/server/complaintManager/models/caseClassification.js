import models from "./index";

module.exports = (sequelize, DataTypes) => {
  const CaseClassification = sequelize.define(
    "case_classification",
    {
      caseId: {
        allowNull: false,
        type: DataTypes.INTEGER,
        field: "case_id",
        references: {
          model: models.cases,
          key: "id"
        }
      },
      classificationId: {
        allowNull: false,
        type: DataTypes.INTEGER,
        field: "classification_id",
        references: {
          model: models.classification,
          key: "id"
        }
      },
      createdAt: {
        allowNull: false,
        type: DataTypes.DATE,
        field: "created_at"
      },
      updatedAt: {
        allowNull: false,
        type: DataTypes.DATE,
        field: "updated_at"
      },
      deletedAt: {
        type: DataTypes.DATE,
        field: "deleted_at"
      }
    },
    { tableName: "case_classifications", paranoid: true }
  );

  CaseClassification.associate = models => {
    CaseClassification.belongsTo(models.classification, {
      as: "classification",
      foreignKey: {
        name: "classificationId",
        field: "classification_id",
        allowNull: false
      }
    });
  };

  CaseClassification.prototype.getCaseId = async function(transaction) {
    return this.caseId;
  };

  CaseClassification.prototype.modelDescription = async function(transaction) {
    const classification = await sequelize
      .model("classification")
      .findByPk(this.classificationId, {
        transaction
      });
    return [
      { "Classification Name": classification.name },
      { Message: classification.message }
    ];
  };

  CaseClassification.auditDataChange();
  return CaseClassification;
};
