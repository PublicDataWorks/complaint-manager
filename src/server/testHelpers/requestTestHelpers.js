import jwt from "jsonwebtoken";
import fs from "fs";
import path from "path";
import models from "../policeDataManager/models/index";
import winston from "winston";

const config = require("../config/config")[process.env.NODE_ENV];

export const suppressWinstonLogs = test => async () => {
  winston.configure({
    transports: [
      new winston.transports.Console({
        json: config.winston.json,
        colorize: true,
        silent: true
      })
    ],
    level: config.winston.logLevel,
    colorize: true
  });

  try {
    await test();
  } catch (err) {
    throw err;
  } finally {
    winston.configure({
      transports: [
        new winston.transports.Console({
          json: config.winston.json,
          colorize: true
        })
      ],
      level: config.winston.logLevel,
      colorize: true
    });
  }
};

export const buildTokenWithPermissions = (permissions, nickname) => {
  const privateKeyPath = path.join(
    __dirname,
    "../config",
    "test",
    "private.pem"
  );
  const cert = fs.readFileSync(privateKeyPath);

  const payload = {
    foo: "bar",
    scope: `${config.authentication.scope} ${permissions}`
  };
  payload[`${config.authentication.nicknameKey}`] = nickname;

  const options = {
    audience: config.authentication.audience,
    issuer: config.authentication.issuer,
    algorithm: config.authentication.algorithm
  };

  return jwt.sign(payload, cert, options);
};

export const expectResponse = async (
  responsePromise,
  statusCode,
  responseBodyContents = null
) => {
  return await responsePromise.then(
    response => {
      expect(response.statusCode).toEqual(statusCode);
      responseBodyContents &&
        expect(response.body).toEqual(responseBodyContents);
      return response;
    },
    error => {
      expect(error.response.statusCode).toEqual(statusCode);
      responseBodyContents &&
        expect(error.response.body).toEqual(responseBodyContents);
      return error;
    }
  );
};

export const cleanupDatabase = async () => {
  const truncationQuery =
    "TRUNCATE referral_letter_officer_recommended_actions CASCADE;" +
    "TRUNCATE referral_letter_officer_history_notes CASCADE;" +
    "TRUNCATE letter_officers CASCADE;" +
    "TRUNCATE referral_letters CASCADE;" +
    "TRUNCATE recommended_actions CASCADE;" +
    "TRUNCATE addresses CASCADE;" +
    "TRUNCATE cases_officers CASCADE;" +
    "TRUNCATE officers_allegations CASCADE;" +
    "TRUNCATE officers CASCADE;" +
    "TRUNCATE allegations CASCADE;" +
    "TRUNCATE classifications CASCADE;" +
    "TRUNCATE civilians CASCADE;" +
    "TRUNCATE attachments CASCADE;" +
    "TRUNCATE case_notes CASCADE;" +
    "TRUNCATE action_audits CASCADE;" +
    "TRUNCATE legacy_data_change_audits CASCADE;" +
    "TRUNCATE intake_sources CASCADE;" +
    "TRUNCATE how_did_you_hear_about_us_sources CASCADE;" +
    "TRUNCATE race_ethnicities CASCADE;" +
    "TRUNCATE officer_history_options CASCADE;" +
    "TRUNCATE cases CASCADE;" +
    "TRUNCATE gender_identities CASCADE;" +
    "TRUNCATE case_note_actions CASCADE;" +
    "TRUNCATE audits CASCADE;" +
    "TRUNCATE case_tags CASCADE;" +
    "TRUNCATE tags CASCADE;" +
    "TRUNCATE civilian_titles CASCADE;" +
    "TRUNCATE districts CASCADE;" +
    "TRUNCATE case_classifications CASCADE;" +
    "TRUNCATE notifications CASCADE;";

  await models.sequelize.query(truncationQuery, {
    type: models.sequelize.QueryTypes.RAW
  });
};
