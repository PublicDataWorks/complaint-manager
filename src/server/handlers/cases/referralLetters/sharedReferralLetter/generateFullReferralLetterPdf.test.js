import { generateLetterPdfHtml } from "./generateFullReferralLetterPdf";
import timekeeper from "timekeeper";
import Case from "../../../../../client/testUtilities/case";
import ReferralLetter from "../../../../../client/testUtilities/ReferralLetter";
import { cleanupDatabase } from "../../../../testHelpers/requestTestHelpers";
import generateReferralLetterFromCaseData from "../generateReferralLetterFromCaseData";
import models from "../../../../models";
import { CASE_STATUS } from "../../../../../sharedUtilities/constants";
import generateFullReferralLetterPdf from "./generateFullReferralLetterPdf";

jest.mock("../generateReferralLetterFromCaseData");

describe("generateFullReferralLetterPdf", () => {
  let existingCase, referralLetter;

  afterEach(async () => {
    await cleanupDatabase();
    generateReferralLetterFromCaseData.mockReset();
  });

  beforeEach(async () => {
    const caseAttributes = new Case.Builder()
      .defaultCase()
      .withId(12070)
      .withFirstContactDate("2017-12-25");
    existingCase = await models.cases.create(caseAttributes, {
      auditUser: "test"
    });
    await existingCase.update(
      { status: CASE_STATUS.ACTIVE },
      { auditUser: "test" }
    );
    await existingCase.update(
      { status: CASE_STATUS.LETTER_IN_PROGRESS },
      { auditUser: "test" }
    );

    const referralLetterAttributes = new ReferralLetter.Builder()
      .defaultReferralLetter()
      .withId(undefined)
      .withCaseId(existingCase.id)
      .withRecipient("recipient address")
      .withSender("sender address")
      .withTranscribedBy("transcriber")
      .withIncludeRetaliationConcerns(true);

    referralLetter = await models.referral_letter.create(
      referralLetterAttributes,
      {
        auditUser: "test"
      }
    );
  });

  test("generates letter pdf html correctly", async () => {
    const timeOfDownload = new Date("2018-07-01 19:00:22 CDT");
    timekeeper.freeze(timeOfDownload);
    const letterBody = "<p> Letter Body </p>";
    const pdfData = {
      referralLetter: {
        recipient: "Recipient Address",
        sender: "Sender Address\n Sender Address Second Line",
        transcribedBy: "Transcriber"
      },
      complaintType: "person",
      incidentDate: "2011-04-09"
    };

    const letterPdfHtml = await generateLetterPdfHtml(
      letterBody,
      pdfData,
      existingCase.id
    );
    expect(letterPdfHtml).toMatchSnapshot();
  });

  test("unedited letter generates pdf from case data", async () => {
    await models.sequelize.transaction(async transaction => {
      await generateFullReferralLetterPdf(existingCase.id, transaction);
      expect(generateReferralLetterFromCaseData).toHaveBeenCalledWith(
        existingCase.id,
        transaction
      );
    });
  });

  test("edited letter generates pdf from saved data", async () => {
    await referralLetter.update(
      {
        editedLetterHtml: "<p> edited </p>"
      },
      { auditUser: "test" }
    );
    await models.sequelize.transaction(async transaction => {
      await generateFullReferralLetterPdf(existingCase.id, transaction);
    });
    expect(generateReferralLetterFromCaseData).not.toHaveBeenCalled();
  });
});
