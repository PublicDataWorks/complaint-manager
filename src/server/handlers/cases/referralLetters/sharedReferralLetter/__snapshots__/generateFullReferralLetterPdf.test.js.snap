// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateFullReferralLetterPdf generates letter pdf html correctly 1`] = `
"<html>
  <head>
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"generatePdf.css\\">
  </head>
  <body>
    <div id=\\"pageHeader-first\\">
      <div style=\\"text-align: center;\\">
        <img style=\\"max-width: 223px\\" src=file:/app/src/server/handlers/cases/referralLetters/generatePdf/assets/header_text.png />
      </div>
      <div style=\\"text-align: center; margin-top: 8px\\">
        <img style=\\"max-width: 42px\\" src=file:/app/src/server/handlers/cases/referralLetters/generatePdf/assets/icon.ico />
      </div>
    </div>
    <div id=\\"pageHeader\\" style=\\"font-size:8.5pt; color: #7F7F7F;\\">
      Recipient Address<br/>
      July 01, 2018<br/>
      Page {{page}}
    </div>

    <div id=\\"pageFooter\\" style=\\"text-align: center; margin-top: 16px\\">
            <span  style=\\"display:inline-block; margin: 6px 16px 0 0\\">
                <img style=\\"max-width: 30px\\" src=file:/app/src/server/handlers/cases/referralLetters/generatePdf/assets/icon.ico />

            </span>
      <span style=\\"display:inline-block; font-size:7pt; color: #7F7F7F;\\">
                INDEPENDENT POLICE MONITOR <br />
                2714 Canal Street, Suite 201 | NEW ORLEANS, LOUISIANA | 70119 <br />
                Phone (504) 309-9799| Fax (504) 309-7345
            </span>
      <span  style=\\"display:inline-block; width: 46px\\">&nbsp;</span>
    </div>
    <p style=\\"color: #7F7F7F;\\">
      SUSAN HUTSON
      <br>
      INDEPENDENT POLICE MONITOR
    </p>
    <p><br></p>
    July 01, 2018
    <p><br></p>
    Recipient Address
    <p><br></p>
    <p><strong>RE: Complaint Referral; IPM Complaint -2011-12070</strong></p>
    <p><br></p>
    <p>Dear Recipient Address:</p>
    <p><br></p>
    <p> Letter Body </p>
    <p><br></p>
    Sincerely,
    <p><br></p>
    <p><br></p>
    <p><br></p>
    Sender Address<br> Sender Address Second Line
    <p><br></p>
    Transcribed by: Transcriber
  </body>
</html>"
`;
